tests/storages_tests/rdb_tests/test_storage.py

```diff
@@ -286,3 +301,60 @@ def test_get_trials_excluded_trial_ids() -> None:
     # See https://github.com/optuna/optuna/issues/1457.
     trials = storage._get_trials(study_id, states=None, excluded_trial_ids=set(range(500000)))
     assert len(trials) == 0
+
+
+def test_record_heartbeat() -> None:
```

Comment:
Maybe I'm doing something strange but I'm seeing errors from the spawned thread that the table doesn't exists. Have you encountered this before?

```
> pytest tests/storages_tests/rdb_tests/test_storage.py -x -s -k 'test_record_heartbeat'
===================================================================================== test session starts =====================================================================================
platform darwin -- Python 3.7.6, pytest-6.2.1, py-1.9.0, pluggy-0.13.1
rootdir: /Users/hvy/repos/optuna, configfile: pytest.ini
plugins: repeat-0.8.0, nbval-0.9.6, flaky-3.6.1
collected 27 items / 26 deselected / 1 selected

tests/storages_tests/rdb_tests/test_storage.py [I 2021-01-20 16:18:30,908] A new study created in RDB with name: no-name-1bfb2d33-1215-4cf2-86e4-f03efb262e45
Exception in thread Thread-1:
Traceback (most recent call last):
  File "/Users/hvy/.pyenv/versions/3.7.6/envs/3.7.6-optuna/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 1246, in _execute_context
    cursor, statement, parameters, context
  File "/Users/hvy/.pyenv/versions/3.7.6/envs/3.7.6-optuna/lib/python3.7/site-packages/sqlalchemy/engine/default.py", line 588, in do_execute
    cursor.execute(statement, parameters)
sqlite3.OperationalError: no such table: trial_heartbeats

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/hvy/repos/optuna/optuna/storages/_rdb/storage.py", line 52, in _create_scoped_session
    yield session
  File "/Users/hvy/repos/optuna/optuna/storages/_rdb/storage.py", line 1150, in record_heartbeat
    heartbeat = models.TrialHeartbeatModel.where_trial_id(trial_id, session)
  File "/Users/hvy/repos/optuna/optuna/storages/_rdb/models.py", line 507, in where_trial_id
    return session.query(cls).filter(cls.trial_id == trial_id).one_or_none()
  File "/Users/hvy/.pyenv/versions/3.7.6/envs/3.7.6-optuna/lib/python3.7/site-packages/sqlalchemy/orm/query.py", line 3316, in one_or_none
    ret = list(self)
  File "/Users/hvy/.pyenv/versions/3.7.6/envs/3.7.6-optuna/lib/python3.7/site-packages/sqlalchemy/orm/query.py", line 3389, in __iter__
    return self._execute_and_instances(context)
  File "/Users/hvy/.pyenv/versions/3.7.6/envs/3.7.6-optuna/lib/python3.7/site-packages/sqlalchemy/orm/query.py", line 3414, in _execute_and_instances
    result = conn.execute(querycontext.statement, self._params)
  File "/Users/hvy/.pyenv/versions/3.7.6/envs/3.7.6-optuna/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 982, in execute
    return meth(self, multiparams, params)
  File "/Users/hvy/.pyenv/versions/3.7.6/envs/3.7.6-optuna/lib/python3.7/site-packages/sqlalchemy/sql/elements.py", line 293, in _execute_on_connection
    return connection._execute_clauseelement(self, multiparams, params)
  File "/Users/hvy/.pyenv/versions/3.7.6/envs/3.7.6-optuna/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 1101, in _execute_clauseelement
    distilled_params,
  File "/Users/hvy/.pyenv/versions/3.7.6/envs/3.7.6-optuna/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 1250, in _execute_context
    e, statement, parameters, cursor, context
  File "/Users/hvy/.pyenv/versions/3.7.6/envs/3.7.6-optuna/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 1476, in _handle_dbapi_exception
    util.raise_from_cause(sqlalchemy_exception, exc_info)
  File "/Users/hvy/.pyenv/versions/3.7.6/envs/3.7.6-optuna/lib/python3.7/site-packages/sqlalchemy/util/compat.py", line 398, in raise_from_cause
    reraise(type(exception), exception, tb=exc_tb, cause=cause)
  File "/Users/hvy/.pyenv/versions/3.7.6/envs/3.7.6-optuna/lib/python3.7/site-packages/sqlalchemy/util/compat.py", line 152, in reraise
    raise value.with_traceback(tb)
  File "/Users/hvy/.pyenv/versions/3.7.6/envs/3.7.6-optuna/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 1246, in _execute_context
    cursor, statement, parameters, context
  File "/Users/hvy/.pyenv/versions/3.7.6/envs/3.7.6-optuna/lib/python3.7/site-packages/sqlalchemy/engine/default.py", line 588, in do_execute
    cursor.execute(statement, parameters)
sqlalchemy.exc.OperationalError: (sqlite3.OperationalError) no such table: trial_heartbeats
[SQL: SELECT trial_heartbeats.trial_heartbeat_id AS trial_heartbeats_trial_heartbeat_id, trial_heartbeats.trial_id AS trial_heartbeats_trial_id, trial_heartbeats.heartbeat AS trial_heartbeats_heartbeat
FROM trial_heartbeats
WHERE trial_heartbeats.trial_id = ?]
[parameters: (1,)]
(Background on this error at: http://sqlalche.me/e/e3q8)

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/hvy/.pyenv/versions/3.7.6/lib/python3.7/threading.py", line 926, in _bootstrap_inner
    self.run()
  File "/Users/hvy/.pyenv/versions/3.7.6/lib/python3.7/threading.py", line 870, in run
    self._target(*self._args, **self._kwargs)
  File "/Users/hvy/repos/optuna/optuna/_optimize.py", line 338, in _record_heartbeat
    storage.record_heartbeat(trial_id)
  File "/Users/hvy/repos/optuna/optuna/storages/_cached_storage.py", line 446, in record_heartbeat
    self._backend.record_heartbeat(trial_id)
  File "/Users/hvy/repos/optuna/optuna/storages/_rdb/storage.py", line 1155, in record_heartbeat
    heartbeat.heartbeat = session.execute(func.now()).scalar()
  File "/Users/hvy/.pyenv/versions/3.7.6/lib/python3.7/contextlib.py", line 130, in __exit__
    self.gen.throw(type, value, traceback)
  File "/Users/hvy/repos/optuna/optuna/storages/_rdb/storage.py", line 70, in _create_scoped_session
    raise optuna.exceptions.StorageInternalError(message) from e
optuna.exceptions.StorageInternalError: An exception is raised during the commit. This typically happens due to invalid data in the commit, e.g. exceeding max length.

[I 2021-01-20 16:18:32,952] Trial 0 finished with value: 1.0 and parameters: {}. Best is trial 0 with value: 1.0.
```

Reply:
Thanks for raising the issue. I confirmed this error is avoidable if we call `models.BaseModel.metadata.create_all(self.engine)` at the head of `RDBStorage.record_heartbeat`. I'm not sure, but the tables are created only in the main thread by calling `models.BaseModel.metadata.create_all(self.engine)` in `RDBStorage.__init__` and are not available in the spawned threads (I checked that other tables such as `trials` or `trial_values` are not available in the spawned threads). 

Please see https://github.com/optuna/optuna/pull/2190/commits/4d120c9e1b955677d15de48aba2064be36f8d8da.

Reply:
Ah, seems like it's specific to SQLite with its in-memory (i.e. `sqlite:///:memory:`). It's an issue with `Study.optimize(..., n_jobs=2)` as well 🤔